kotlinx-collections-immutable/tests/src/test/kotlin/kotlinx.collections.immutable/implementations/immutableMap/PersistentHashMapTest.kt

```diff
@@ -0,0 +1,362 @@
+
+        val elementsToAdd = 1000000
+        repeat(times = elementsToAdd) { index ->
+            map = map.put(index, index.toString())
+        }
+        repeat(times = elementsToAdd) { index ->
+            assertEquals(elementsToAdd - index, map.size)
+            assertEquals(index.toString(), map[index])
+            map = map.remove(index)
+            assertNull(map[index])
+        }
+    }
+
+    @Test
+    fun removeEntryTests() {
+        var map = persistentHashMapOf<Int, String>()
+        assertTrue(map.put(0, "0").remove(0, "0").isEmpty())
+        assertFalse(map.put(0, "0").remove(0, "x").isEmpty())
+
+        val elementsToAdd = 1000000
+        repeat(times = elementsToAdd) { index ->
+            map = map.put(index, index.toString())
+        }
+        repeat(times = elementsToAdd) { index ->
+            assertEquals(elementsToAdd - index, map.size)
+            assertEquals(index.toString(), map[index])
+            map = map.remove(index, (index + 1).toString())
+            assertEquals(index.toString(), map[index])
+            map = map.remove(index, index.toString())
+            assertNull(map[index])
+        }
+    }
+
+    @Test
+    fun getTests() {
+        var map = persistentHashMapOf<Int, String>()
+        assertEquals("1", map.put(1, "1")[1])
+
+        val elementsToAdd = 10000
+        repeat(times = elementsToAdd) { index ->
+            map = map.put(index, index.toString())
+
+            for (i in 0..index) {
+                assertEquals(i.toString(), map[i])
+            }
+        }
+        repeat(times = elementsToAdd) { index ->
+            for (i in elementsToAdd - 1 downTo index) {
+                assertEquals(i.toString(), map[i])
+            }
+
+            map = map.remove(index)
+        }
+    }
+
+    @Test
+    fun putTests() {
+        var map = persistentHashMapOf<Int, String>()
+        assertEquals("2", map.put(1, "1").put(1, "2")[1])
+
+        val elementsToAdd = 5000
+        repeat(times = elementsToAdd) { index ->
+            map = map.put(index, (index * 2).toString())
+
+            for (i in 0..index) {
+                assertEquals((i + index).toString(), map[i])
+                map = map.put(i, (i + index + 1).toString())
+                assertEquals((i + index + 1).toString(), map[i])
+            }
+        }
+        repeat(times = elementsToAdd) { index ->
+            for (i in index until elementsToAdd) {
+                val expected = elementsToAdd - index + i
+
+                assertEquals(expected.toString(), map[i])
+                map = map.put(i, (expected - 1).toString())
+                assertEquals((expected - 1).toString(), map[i])
+            }
+
+            map = map.remove(index)
+        }
+    }
+
+    @Test
+    fun collisionTests() {
+        var map = persistentHashMapOf<KeyWrapper<Int>, Int>()
+
+        assertEquals(2, map.put(KeyWrapper(1, 1), 1).put(KeyWrapper(1, 1), 2)[KeyWrapper(1, 1)])
```

Comment:
This isn't a collision, it's just a value update for a given key.

Reply:
+1